### PR TITLE
Fix bug in MarkdownHeaderTextSplitter for codeblock

### PR DESCRIPTION
**Pull request by Dynmi**
- Description: The previous version of the MarkdownHeaderTextSplitter did not take into account the possibility of '#' appearing within code blocks, which caused segmentation anomalies in these situations. This PR has fixed this issue.
  - Issue: 
  - Dependencies: No
  - Tag maintainer: 
  - Twitter handle: 

cc @baskaryan @eyurtsev  @rlancemartin

### CODE REVIEW

**Comment by baskaryan** [comment]
hey @Dynmi! thanks for the PR, could we add some unit tests?

**Comment by Dynmi** [comment]
ok, updated just now. @baskaryan

**Comment by Dynmi** [comment]
Please review the PR again. @baskaryan  thanks:)

File: libs/langchain/langchain/text_splitter.py [hunk 1]
```diff
@@ -382,8 +382,25 @@ def split_text(self, text: str) -> List[Document]:
         header_stack: List[HeaderType] = []
         initial_metadata: Dict[str, str] = {}
 
+        in_code_block = False
+
         for line in lines:
             stripped_line = line.strip()
+
+            if stripped_line.startswith("```"):
+                if not in_code_block:  # code block starts
```
Comment (eyurtsev) [inline libs/langchain/langchain/text_splitter.py:391]: The if statements can be replaced with

```python
in_code_block = not in_code_block
```

**Review by eyurtsev — COMMENTED** [review]
@baskaryan -- I'm not exactly sure what the markdown header text splitter is supposed to do -- behavior isn't documented. 

The content of the code block is missing from the chunks

**Comment by Dynmi** [comment]
I think that ‘commit#8c6be78f17a52c28aa77f50ed9c41c3931eca827’ can be merged into master? @baskaryan

**Comment by Dynmi** [comment]
Could someone take a look at this pull request please? @baskaryan @eyurtsev @rlancemartin

**Comment by Dynmi** [comment]
Could someone take a look at this pull request please? @baskaryan @eyurtsev @rlancemartin

**Comment by Dynmi** [comment]
Could someone take a look at this pull request please? 

@baskaryan @eyurtsev @rlancemartin @hwchase17

**Comment by baskaryan** [comment]
looks reasonable to me, sorry for delay and thanks for the improvement @Dynmi!

**Pull request merged**